diff -Naur opkg-utils-r5966/makePackage opkg-utils-r5966-mod/makePackage
--- opkg-utils-r5966/makePackage	2010-05-25 15:18:38.000000000 +0300
+++ opkg-utils-r5966-mod/makePackage	2012-02-28 10:46:30.546101365 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python2
 
 # The general algorithm this program follows goes like this:
 #   Run tar to extract control from control.tar.gz from the package.
diff -Naur opkg-utils-r5966/opkg-list-fields opkg-utils-r5966-mod/opkg-list-fields
--- opkg-utils-r5966/opkg-list-fields	2010-05-25 15:18:38.000000000 +0300
+++ opkg-utils-r5966-mod/opkg-list-fields	2012-02-28 10:45:58.189436081 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python2
 
 import sys, opkg
 
diff -Naur opkg-utils-r5966/opkg-make-index opkg-utils-r5966-mod/opkg-make-index
--- opkg-utils-r5966/opkg-make-index	2010-05-25 15:18:38.000000000 +0300
+++ opkg-utils-r5966-mod/opkg-make-index	2012-02-28 10:46:03.776102509 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python2
 
 import sys, os, posixpath
 from glob import glob
diff -Naur opkg-utils-r5966/opkg-show-deps opkg-utils-r5966-mod/opkg-show-deps
--- opkg-utils-r5966/opkg-show-deps	2010-05-25 15:18:38.000000000 +0300
+++ opkg-utils-r5966-mod/opkg-show-deps	2012-02-28 10:46:08.019435661 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python2
 
 import sys, os, posixpath
 from glob import glob
diff -Naur opkg-utils-r5966/opkg-unbuild opkg-utils-r5966-mod/opkg-unbuild
--- opkg-utils-r5966/opkg-unbuild	2010-05-25 15:18:38.000000000 +0300
+++ opkg-utils-r5966-mod/opkg-unbuild	2012-02-28 10:46:14.796102038 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python2
 
 import sys, os, re
 
